DOC Extending the vocab and storing embeddings (#1335)
huggingface/peft: docs/source/developer_guides/troubleshooting.md

```

The mentioned layers should be added to `modules_to_save` in the config to avoid the described problem.

### Extending the vocabulary

> For many language fine-tuning tasks, it's necessary to extend the vocabulary of the model, since new tokens are being introduced. When this is done, the embedding layer needs to be extended to account for the new tokens. But this means that the embedding layer should be stored in addition to the adapter weights when saving the adapter.

Reviewer: ```suggestion
For many language fine-tuning tasks, extending the model's vocabulary is necessary since new tokens are being introduced. This requires extending the embedding layer to account for the new tokens and also storing the embedding layer in addition to the adapter weights when saving the adapter.
```

> To ensure that this works, add the embedding layer to the `target_modules` of the config. As an example, for Mistral, the config could look like this:

Reviewer: ```suggestion
Save the embedding layer by adding it to the `target_modules` of the config. The embedding layer name must follow the standard naming scheme from Transformers. For example, the Mistral config could look like this:
```

```python
config = LoraConfig(..., target_modules=["embed_tokens", "lm_head", "q_proj", "v_proj"])
```

> Then, PEFT will automatically take care of storing the embedding layer when saving the adapter. However, for this to work, the name of the embedding layer needs to follow the standard naming scheme of 🤗 Transformers. If it doesn't, you can manually pass `save_embedding_layers=True` when saving the adapter:

Reviewer: ```suggestion
Once added to `target_modules`, PEFT automatically stores the embedding layer when saving the adapter if the model has the [`~transformers.PreTrainedModel.get_input_embeddings`] and [`~transformers.PreTrainedModel.get_output_embeddings`]. This is generally the case for Transformers models.

If the model's embedding layer doesn't follow the Transformer's naming scheme, you can still save it by manually passing `save_embedding_layers=True` when saving the adapter:
```

```python
model = get_peft_model(...)
# train the model
model.save_adapter("my_adapter", save_embedding_layers=True)
```

> This requires the model to have the methods `get_input_embeddings` and `get_output_embeddings`, which is generally the case for 🤗 Transformers models.

Reviewer: ```suggestion
```

> For inference, please ensure that you first load the base model, then resize it in the same way as you did before you trained the model, and only then load the PEFT checkpoint.

Reviewer: ```suggestion
For inference, load the base model first and resize it the same way you did before you trained the model. After you've resized the base model, you can load the PEFT checkpoint.
```

Save the embedding layer by adding it to the `target_modules` of the config. The embedding layer name must follow the standard naming scheme from Transformers. For example, the Mistral config could look like this:

```python
config = LoraConfig(..., target_modules=["embed_tokens", "lm_head", "q_proj", "v_proj"])
```

Once added to `target_modules`, PEFT automatically stores the embedding layer when saving the adapter if the model has the [`~transformers.PreTrainedModel.get_input_embeddings`] and [`~transformers.PreTrainedModel.get_output_embeddings`]. This is generally the case for Transformers models.

If the model's embedding layer doesn't follow the Transformer's naming scheme, you can still save it by manually passing `save_embedding_layers=True` when saving the adapter:

```python
model = get_peft_model(...)
# train the model
model.save_adapter("my_adapter", save_embedding_layers=True)
```

> 

Reviewer 2: ```suggestion
```
